File: RssReaderObj/RssReaderObj/RssParser.m
//
//  RssParser.m
//  RssReaderObj
//
//  Created by Admin on 11.10.16.
//  Copyright © 2016 Admin. All rights reserved.
//

#import "RssParser.h"
#import "RssItem.h"

@implementation RssParser
NSString *rssUrl = @"http://www.popmech.ru/out/public-all.xml";
NSMutableString *eName;

RssItem *currentItem;

-(instancetype)init
{
    self = [super init];
    if (self)
    {
       
        self.feeds = [NSMutableArray arrayWithCapacity:1];
        self.parser = [[NSXMLParser alloc]initWithContentsOfURL:[NSURL URLWithString:rssUrl ]];
        self.parser.delegate = self;
    }
    return self;
}
-(BFTask *)parse
{
    BFTaskCompletionSource *tsc = [BFTaskCompletionSource taskCompletionSource];
    BOOL success = [self.parser parse];
    if (success)
    {
        //if (_delegate != NULL)
        //{
            //[_delegate setParsedData:_feeds];
            [tsc setResult:_feeds];
       // }
        for (RssItem *f in _feeds)
        {
            [f print];
        }
    }
    else
    {  [tsc setResult:nil];
    }    return tsc.task;
}
- (void)parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName attributes:(NSDictionary *)attributeDict{

    eName = [NSMutableString stringWithString:elementName];
    if ([elementName  isEqual: @"item"])
    {
        currentItem = [[RssItem alloc]init];
    }
    else
    {
        if ([elementName  isEqual: @"enclosure"])
        {
            [currentItem.imageUrl appendString:[NSMutableString stringWithString: attributeDict[@"url"]]];
        }
    }
}
- (void)parser:(NSXMLParser *)parser foundCharacters:(NSString *)string{
    NSString *data = [string stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if (![data isEqual:@""])
    {
    if (currentItem != nil)
    {
    if ([eName isEqual:@"title"])
    {
        [currentItem.title appendString:[data copy]];
    }
    else
        if ([eName isEqual:@"pubDate"])
        {
           [currentItem.pubDate appendString:[data copy]];
        }
    else
        if ([eName isEqual:@"link"])
        {
           [currentItem.link appendString:[data copy]];
        }
    }
    }
}
-(void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName
{
    if ([elementName isEqual:@"item"])
    {
        if (currentItem != nil)
        {
        [self.feeds addObject:[currentItem  createcopy]];
        }
    }
}

@end

File: MechanicReader/MechanicReader/RssParser.swift
//
//  RssParser.swift
//  MechanicReader
//
//  Created by Admin on 10.10.16.
//  Copyright © 2016 Admin. All rights reserved.
//

import Foundation
import BoltsSwift

public class RssParser:NSObject, NSXMLParserDelegate
{
    private  let rssURL = "http://www.popmech.ru/out/public-all.xml"
    private var parser:NSXMLParser?
    
    var feed=[RssItem]()
    private var currentElement:RssItem?
    private var eName:String=""
    
   public override init() {
        self.parser = NSXMLParser()
   
        super.init()
     self.parser?.delegate = self 
    self.setParser()
    }
    private func setParser()
    {
       self.parser = NSXMLParser(contentsOfURL: NSURL(string: rssURL)!)
        self.parser?.delegate = self
    }
    public func parse()->Task<Bool>
    {
        let tcs = TaskCompletionSource<Bool>()
        let queue = dispatch_queue_create("rss.parser", DISPATCH_QUEUE_CONCURRENT)
        dispatch_async(queue){
        if let success = self.parser?.parse()
        {
            dispatch_async(dispatch_get_main_queue())
            {
            tcs.set(result: success)
            }
            
            
        
        }
        else
        {
            dispatch_async(dispatch_get_main_queue())
            {
                tcs.set(result: false)
            }
        }
        }
    
        return tcs.task
    }
    
    
    public func parser(parser: NSXMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String : String]) {
        self.eName = elementName
        
        if eName == "item"
        {
            self.currentElement = RssItem()
        }
        else {
            if eName == "enclosure"
            {
                if let url = attributeDict["url"]
                {
                self.currentElement?.imageUrl = url
                }
            }
        }
    }
    public func parser(parser: NSXMLParser, foundCharacters string: String) {
        let data = string.stringByTrimmingCharactersInSet(NSCharacterSet.whitespaceAndNewlineCharacterSet())
        if !data.isEmpty
        {
            if eName == "title"
            {
                self.currentElement?.title += data
            }
            else
                if eName == "link"
                {
                    self.currentElement?.link += data
                }
                    
                else if eName == "pubDate"
                {
                    self.currentElement?.pubDate += data
            }
            
        }
    }
    public func parser(parser: NSXMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        if elementName == "item"
        {
            if let el = self.currentElement
            {
                self.feed.append(el)
            }
        }
    }
}